File: 90-SubsetsIi/90-SubsetsIi.java
// Last updated: 8/16/2026, 9:52:41 PM
class Solution {
    // public List<List<Integer>> subsetsWithDup(int[] nums) {
    //     // []
    //     // [] + [1]
    //     // [], [1] + [2`], [1, 2`]
    //     // [], [1], [2`], [1, 2`] + [2``], [1, 2``], [2`, 2``], [1, 2`, 2``]
    //     //                          skip   skip      keep      keep
    //     // for 2`` loop of 'for(List<Integer> curr : res)' 
    //     // -> skip first 2 subsets (results from nums[0]'s loop), start from [2`], [1, 2`]

    //     List<List<Integer>> res = new ArrayList<>();
    //     res.add(new ArrayList<Integer>());
    //     Arrays.sort(nums);

    //     int cachedSize = 0, startIdx = 0;
    //     for(int i = 0; i < nums.length; i ++) {
    //         List<List<Integer>> subsets = new ArrayList<>();
    //         // set startIdx before updating cachedSize
    //         // startIdx = size of previous res size
    //         startIdx = (i > 0 && nums[i] == nums[i-1]) ? cachedSize : 0;
    //         cachedSize = res.size();

    //         for(int j = startIdx; j < res.size(); j ++) {
    //             // add 'nums[i]' to each subset under 'rst'
    //             // make a deep copy of curr, 
    //             // otherwise if you add value to 'curr', the 'rst' lists will change too
    //             List<Integer> curr = res.get(j);
    //             List<Integer> temp = new ArrayList<>(curr); // deep copy
    //             temp.add(nums[i]);
    //             subsets.add(temp);
    //         }
    //         // add 'subsets' results into 'rst'
    //         // for(List<Integer> subset : subsets)
    //         //     res.add(subset);
    //         res.addAll(subsets);
    //     }
    //     return res;
    // }

    // method 2 - DFS
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        helper(res, new ArrayList<Integer>(), nums, 0);
        return res;
    }

    // return all subsets starting with 'subset', next avaible number from 'startIdx'
    // subset = [1], startIdx = 1 ->
    // [1] -> [1, 2`], [1, 2`, 2``], [1, 2``]
    // [2`] -> [2`, 2``]
    // [2``] 
    // discard [1, 2``], [2``]
    private void helper(List<List<Integer>> res, List<Integer> subset, int[] nums, int startIdx) {
        // res.add(subset); wrong, should deep copy 'subset'. otherwise 'subset will change afterwards
        res.add(new ArrayList<>(subset));

        System.out.println(subset);
        for(int i = startIdx; i < nums.length; i ++) {
            if(i > startIdx && nums[i] == nums[i - 1]) continue;
            // (i > 0 && nums[i] == nums[i - 1]) is wrong
            // cause for [1,2`,2``], it will skip the 2`` selection from {subset=[1,2`], startIdx=2}
            // so the skip should be: skip NEXT same value candidate
            // do not skip CURRENT candidate
            subset.add(nums[i]);
            // System.out.println(subset);
            helper(res, subset, nums, i + 1);
            subset.remove(subset.size() - 1);
        }

    }

}

File: 90-SubsetsII/90-SubsetsII.py
# Last updated: 8/23/2026, 12:58:05 PM
1class Solution:
2    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
3        # sort nums, and pick next DIFF num (skip same values)
4        # backtrack
5
6        nums.sort()
7        res = []
8        path = []   # curr path
9
10        # find legit subset in nums[start:]
11        def backtrack(start: int):
12            res.append(path[:]) # copy path as path is changing
13
14            for i in range(start, len(nums)):
15                # i > start not i > 0
16                # as i==start is the first choice, it should always be selected, otherwise you will skip legit combinations
17                # 去重核心：i > start（不是 i > 0）——只禁止"同一层for循环里横向选到相同数字"
18                # 只有i > start时，才是"在同一层里，前一个兄弟分支已经试过某个值了，现在轮到试下一个候选值"，这时才需要跟"同一层的上一个候选值"比较去重。
19                # 允许"同一条路径纵向连续选相同数字"（比如[2,2]这个子集本身是合法的，必须能生成）
20                if i > start and nums[i] == nums[i-1]:
21                    continue
22                
23                path.append(nums[i])    # select
24                backtrack(i + 1)        # check next element
25                path.pop()              # cancel selection
26        
27        backtrack(0)
28        return res
29
30